src/identity/auth.ts

```diff
@@ -0,0 +1,135 @@
+import {NextFunction, Request, Response, Handler} from 'express'
+import * as log4js from 'log4js'
+import {PassportStatic} from 'passport'
+import {IdentityService} from './identityService'
+import * as oauth2 from 'passport-oauth2'
+import {Identity} from './identity'
+
+const logger = log4js.getLogger('config/passport')
+
+export class Auth {
+	public static get AUTHENTICATION_PATH(): string {
+		return '/authenticate'
+	}
+	public static get REDIRECT_COOKIE_NAME(): string {
+		return 'redirectTo'
+	}
+
+	clientId: string
+	clientSecret: string
+	authenticationServiceUrl: string
+	callbackUrl: string
+	passportStatic: PassportStatic
+	identityService: IdentityService
+
+	constructor(
+		clientId: string,
+		clientSecret: string,
+		authenticationServiceUrl: string,
+		callbackUrl: string,
+		passportStatic: PassportStatic,
+		identityService: IdentityService
+	) {
+		this.clientId = clientId
+		this.clientSecret = clientSecret
+		this.authenticationServiceUrl = authenticationServiceUrl
+		this.callbackUrl = callbackUrl
+		this.passportStatic = passportStatic
+		this.identityService = identityService
+	}
+
+	initialize(): Handler {
+		return this.passportStatic.initialize()
+	}
+
+	session(): Handler {
+		return this.passportStatic.session()
+	}
+
+	configureStrategy() {
+		let strategy: oauth2.Strategy
+		strategy = new oauth2.Strategy(
+			{
+				authorizationURL: `${this.authenticationServiceUrl}/oauth/authorize`,
+				callbackURL: `${this.callbackUrl}/authenticate`,
+				clientID: this.clientId,
+				clientSecret: this.clientSecret,
+				tokenURL: `${this.authenticationServiceUrl}/oauth/token`,
+			},
+			this.verify()
+		)
+		this.passportStatic.use(strategy)
+
+		this.passportStatic.serializeUser((user, done) => {
+			done(null, JSON.stringify(user))
+		})
+
+		this.passportStatic.deserializeUser<Identity, string>(this.deserializeUser())
+	}
+
+	verify() {
+		return async (
+			accessToken: string,
+			refreshToken: string,
+			profile: any,
+			cb: oauth2.VerifyCallback
+		) => {
+			try {
+				const identityDetails = await this.identityService.getDetails(
+					accessToken
+				)
+
+				cb(null, identityDetails)
+			} catch (e) {
+				logger.warn(`Error retrieving user profile information`, e)
+				cb(e)
+			}
+		}
+	}
+
+	checkAuthenticated() {
+		return (req: Request, res: Response, next: NextFunction) => {
+			if (req.isAuthenticated()) {
+				return next()
+			}
+
+			res.cookie(Auth.REDIRECT_COOKIE_NAME, req.originalUrl)
+			res.redirect(Auth.AUTHENTICATION_PATH)
+		}
+	}
+
+	authenticate() {
+		return this.passportStatic.authenticate('oauth2', {
+			failureFlash: true,
+			failureRedirect: '/',
+		})
+	}
+
+	redirect() {
+		return (req: Request, res: Response) => {
+			const redirect = req.cookies[Auth.REDIRECT_COOKIE_NAME]
+			if (!redirect) {
+				logger.info('passport: session not present on express request')
+				res.sendStatus(500)
```

Comment:
We shouldn't fail here (if it's an error all we've done is forgotten where the user was originally going) - default to redirecting to the home page.